CIPs/0020.md

```diff
@@ -0,0 +1,127 @@
+# CIP [0020]: Generic hash function support via extensible precompile
+
+- Date: 2020-09-4
+- Author: @prestwich (James Prestwich)
+- Status: WIP
+
+### Terminology
+
+The key words "MUST", "MUST NOT", "SHOULD", and "MAY" in this document are to be interpreted as described in 
+[RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html).
+
+## Overview
+
+As a descendant of the EVM, Celo supports only a handful of common hash 
+functions. This restricts the design-space of Celo protocols, and 
+
+## Goals
+
+- Provide support for common cryptographic hash functions
+- Expand options for on-chain verification of protocols not specifically 
+designed for EVM compatibility
+    - e.g. Verification of off-chain signatures using Blake2Xs, including Celo's 
+    own epoch commitments within a smart contract.
+    - e.g. Verification of remote Proof of Work from a wide variety of chains
+- Provide a clear route for acceptance and adoption of additional hash functions
+
+## Proposed Solution
+
+Introduce a new precompiled contract at address [TBD] that provides access to 
+a set of pre-determined hash functions. The input to this pre-compile consists 
+of a one-byte selector, a function-specific parameter block, and the preimage. 
+The output of this pre-compile is dependent on the selector, and may be fixed- 
+or variable-length. If the selector is unknown (not on the current 
+compatibility list), or the function-specific, the precompile MUST return an 
+error.
+
+### Proposing an extension
+
+To be eligible for inclusion, a proposed hash function meet the following 
+criteria:
+
+1. There MUST be a standardized specification for the function.
+1. There MUST be an existing high-quality implementation in Go.
+    1. The proposer SHOULD provide an implementation of the precompile 
+    extension.
+1. The function MUST be pure.
+    1. This is intended to disqualify `ethash` and other stateful functions.
+1. The function MUST NOT be memory intensive
```

Comment:
Could we allow memory hard hash functions given that their gas costs are appropriately tuned? (This obviously might result in the de-facto ban of all memory-hard hash functions except with the most trivial parameters)

Reply:
It seems unwise to offer functions that are designed to consume resources in a strictly resource-constrained environment. Do you have a use case in mind?

Reply:
I was thinking about verifying the results of an Ethhash, scrypt, or Argon2 PoW output, mostly with applications such as bridging in mind, but also for PoW as rate limiting for dApps. None of these are a pressing need though, and it's quite likely that the gas costs would make not a concern in reality, so we could leave this intentionally unaddressed for now.

Reply:
Ah, those are specifically and intentionally excluded